functions/calcTajimaD.txt

```diff
@@ -0,0 +1,65 @@
+// Written by Nick Bailey, submitted on 2024-06-13
+// Currently affiliated with CNRS and University of Lyon 1
+// This code is provided in the public domain free of restriction, please credit SLiM-Extras (https://github.com/MesserLab/SLiM-Extras) and myself, Nick Bailey, if you use it
+
+// calculate neutrality test based on the site frequency spectrum, Tajima's D
+// calculations originally in Tajima 1989
+function (float)calcTajimaD(object<Genome> genomes, [No<Mutation> muts = NULL], [Ni$ start = NULL], [Ni$ end = NULL])
+// conduct various checks for proper usage copied straight from calcWattersonTheta()
+{ if (genomes.length() == 0)
+		stop("ERROR (calcTajimaD()): genomes must be non-empty.");
+	if (community.allSpecies.length() > 1)
+	{
+		species = unique(genomes.individual.subpopulation.species, preserveOrder=F);
+		if (species.length() != 1)
+			stop("ERROR (calcTajimaD()): genomes must all belong to the same species.");
+		if (!isNULL(muts))
+			if (!all(muts.mutationType.species == species))
+				stop("ERROR (calcTajimaD()): muts must all belong to the same species as genomes.");
+	}
+	else
+	{
+		species = community.allSpecies;
+	}
+	
+	if (isNULL(muts))
+		muts = species.mutations;
+	
+	// handle windowing
+	if (!isNULL(start) & !isNULL(end))
+	{
+		if (start > end)
+			stop("ERROR (calcTajimaD()): start must be less than or equal to end.");
+		mpos = muts.position;
+		muts = muts[(mpos >= start) & (mpos <= end)];
+		length = end - start + 1;
+	}
+	else if (!isNULL(start) | !isNULL(end))
+	{
+		stop("ERROR (calcTajimaD()): start and end must both be NULL or both be non-NULL.");
+	}
+
+	// narrow down to the mutations that are actually present in the genomes and aren't fixed
+	p = genomes.mutationFrequenciesInGenomes(muts);
+	muts = muts[(p != 0.0) & (p != 1.0)];
+	
+	// do the calculation
+	// Pi and Watterson's theta functions divide by sequence length so this must be undone in Tajima's D
+	// Sequence length is constant (i.e. no missing data or indels) so this can be applied equally over both metrics
+	diff = (calcPiTheta(genomes) - calcWattersonsTheta(genomes)) * (species.chromosome.lastPosition + 1);
+	// calculate standard deviation of covariance of pi and Watterson's theta
+	k = size(muts);
```

Comment:
Isn't it the case that `calcWattersonsTheta( )` is giving us `k/a_1`? It seems more transparent to either (a) calculate watterson's theta directly here like that, or (b) use `calcWattersonsTheta( )` to get the value of `k` here. (This is not obvious because of various possible definitions like "number of mutations" or "number of sites with a mutation", etc.)

Reply:
k here should represent number of sites with a mutation. Not sure what to do here because I see how it's simpler to just calculate k/a_1 and divide by sites (as is done in calcWattersonsTheta) since that information is present already but as in the above comment I need to change this so intervals work properly over calcPiTheta and calcWattersonTheta.

Reply:
ah, I see - perhaps a comment?

Reply:
Since I kept the above interval handling, I have now just left a comment that the variables given also define Watterson's theta.